### ISSUE TITLE
Will it work for RecSys?

### ISSUE THREAD

**Issue by SaiChandra3030** [jrzaurin/pytorch-widedeep#107]
Hello team, i am working with recommendation system, i came across to your repo for recommendation system model building, now i have doubt. will this library work for the recommendation systems or not

I haven't seen any example on the recommendation system in the documentation.

Thanks 
Sai

**Comment by jrzaurin** [comment]
Hi @SaiChandra3030 , thanks for opening the issue

The library was originally inspired by the wide and deep paper which is indeed designed for RecSys, so, it should "_work_". 

Of course all depends what you mean by "work" 🙂, meaning good results are not guaranteed. However, for example, you can pass some categorical cols to the wide component and some continuous cols to the deep component choosing a `TabMlp` tabular model and then the resulting model would be identical to the one described in the original paper.  You can chose transformers for the Deep component, or a custom model for any component (tabular, text and images) and the library will train. 

Another issue is how you want to train. The library broadly supports 2 types of training: 

- Supervised (both frequentist and bayesian via the `Trainer` and the `BayesianTrainer`)
- Self-Supervised pretraining

So if this is enough for you, the library will train. So in summary, the answer is yes, as long as your data can be represented in a way that can be fed to the library, which I think it is not too restrictive. 

Let me know if you need more help 

In case you find it helpful, here is a notebook I wrote a few years ago back when all this was not a library but a fun side project:  
https://github.com/jrzaurin/RecoTour/blob/master/recotour/Chapter14_Wide_Deep_data_preparation.ipynb

**Comment by SaiChandra3030** [comment]
Hello @jrzaurin Thanks, for the great explanation i have gone through the documentation, wide and deep mostly used for the classification, i have done up to the **classification**. the real issue is that **I haven't find the recommendation model like ranking the top products for the customers.**

if you give clear idea on this. it will be more helpful for me.

**Comment by jrzaurin** [comment]
Hey @SaiChandra3030 

okay, so here we are getting into different discussions. One is whether or not the library can be used to build a recommender system, and the answer to that is "yes". Another one is whether the library supports ranking losses or metrics, and to that the answer is "no" (or not yet) 

So let me give you an example of a possible project flow. 

You need to build a recommendations system and you represent your data as triplets `(user feat, item feat, target)`. Target here can be binary (will the user purchase the item or not), multiclass (how will that user review the item) or continuous (what is the users' interest in the item where interest is defined as a continuous target between 0 and 1, for example). You choose to include some implicit negative feedback (for example, for negative samples per positive) so you end up with a training (and potentially validation) dataset that has a 1/4 ratio of purchases/not-purchases. Then you train an algorithm. Up until this point, all is supported by this library. 

Now you want to evaluate your algorithm using a ranking metric (say NDCG) and to that aim you build a testing dataset with a 1/10 ratio of purchase/not-purchases. You run `.predict` and then rank based on predictions and compute the metric. This last final part is not supported by the library "out of the box" (we do not have ranking metrics right now). 

In other words, at the moment we do not support ranking metrics (e.g. NDCG) or ranking losses (for example marging ranking loss) "out of the box" as these need the data prepared in a particular way.  We can indeed consider it and it would actually be a nice addition to the library. So if you have suggestions or want to contribute, let us know!

Hope all this helps

**Comment by SaiChandra3030** [comment]
Thanks for the Great Explanation, I have got it.

here is a small doubt, i am using multiclass classification targets. here i am going with the `predict_prob` function as we know that it will return `probabilities` of the classes. I am using `argsort` method to sort the index and finding the top `probabilities `of the `predict_prob` . CAN I CONSIDER TOP 10 VALUES AS THE RECOMMENDATIONS FOT THE GIVEN INPUT FROM THAT `predict_prob`.


why am I thinking like this means?
if i use `predict` function i will get back `argmax value` as the output class. now, this second highest `argmax value` as the model thinking it might be its target class but it was in second place this will continue until last target class. if i take top 10 it might be top target classes for the input which is returning by the model.

sorry, for asking lot of doubts :(. as a beginner in recsystems i have lot of doubts.

Thanks @jrzaurin

**Comment by SaiChandra3030** [comment]
> Traveling now, I will reply when I have a sec. and no worries, ask as much as you need, see if we can answer 😉

Waiting for your reply @jrzaurin, reply me in your free time 😄

**Comment by SaiChandra3030** [comment]
> Traveling now, I will reply when I have a sec. and no worries, ask as much as you need, see if we can answer 😉

Hello @jrzaurin hope you're doing well, will please reply for the above comment. Thanks

**Comment by jrzaurin** [comment]
Hi @SaiChandra3030 bear in mind we all have jobs and responsibilities in our daily lives, so I will not always be able to reply immediately after your comments. I will always try to find time, but sometimes you will have to wait a bit. 

Anyway, having said that, the answer to your question is, in principle, yes, you can use the probabilities of your items as scores to rank. 

For example, if you are formulating the problem as a 20 class multi-class classification, you can use the predicted probabilities and rank them to select the top 10, that is a valid approach. 

Hope this helps!

**Comment by SaiChandra3030** [comment]
Thank You so much @jrzaurin and sorry for disturbing :( you. I have got all my answers. 

I hope this thread will help future developers too, I have got accuracy around 60% with 420 classes working on the tuning of the models. 

Thanks :)

**Comment by jrzaurin** [comment]
Awesome, let me close this and if you have any other question, keep posting :)
